### ISSUE TITLE
Multiplayer - Performance report for stress test on 1.5.7 (28-40 players on at same time)

### ISSUE THREAD

**Issue by Aledark789** [anegostudios/VintageStory-Issues#1374]
Here's report you asked for Tyron and Radfast.

Game Version: v1.15.7 (Stable)
C# Framework: Mono 6.12.0.122 (tarball Mon Feb 22 17:29:18 UTC 2021)
Server v1.15.7, network v1.15.4, api v1.7.0
**Platform:** Linux Server
**Modded:** Yes
- BetterCratesCC_v1.3.3.zip
- bricklayers-0.5.0.zip
- CarryCapacity-VS1.15.6-v0.6.3.zip
- ExtraChestsCC_v1.4.0.zip
- FarmlandDropsSoil-VS1.15-v1.4.0.zip
- medievalexpansion-3.6.1.zip
- Th3Essentials_1.0.6.zip'
- tradeomat_v1.15.5-1.5.9.zip
- workbenchexpansion

Server specs:
Intel i7-7700K, 2x SSD NVMe 450GB Soft RAID, 
500 mb internet line, 
32GB DDR4 2133MHz
500 GB of cloud storage backup

World file size is about 95 Gigs. Server been active since August 2020 Roughly 1.13.3?

**### Description**
Analysis of server overloads for 28ish to 40 people online on 24 october 2021 16:00 to roughly 20:00. Roughly 10-12 at the end of the file.
Starts doing /debug logticks 300 at 24.10.2021 15:59:04 (see log file link below)

From parsing the file here's the most damaging ones overall for our test. 
Note at some point we were mostly all in the same chunk to take a picture so that helped the server quite a lot by reducing number of chunks loaded. 

1st People were having trouble using the teleport block which doesn't seem to be prioritize at all by the server...several minutes and many werent able to port over even when trying multiple times.
2nd Translocators were also much slower to respond and port you over (roughly 2 minutes) when it did work...
3rd the chunk loading wasnt good at all if people were crossing chunk borders they had to wait for it to render.

The data could of been different if we all went exploring but that could be done on an other time!
Heres the full data : https://gist.githubusercontent.com/Aledark789/97a4af79721c8a7804a821e1a3581327/raw/b776b4c627339d5df9c035f8869e126c15b9f103/gistfile1.txt

**### Analysis of ticks overloads**
Tier 1 culprits (Highest % contributing to the the server overloading and most recurrent ones)
entity-done-bh-bodytemperature (on higher count player this count sometimes for up to 30%-50% of a tick)
entity-controlledphysics-end (Usually least 25%-35% of a whole tick)
entity-ai-tasks-tick-start-execVintagestory.GameContent.AiTaskSeekEntity (when people are split up..its quite extensive).

Tier 2 culprits (Decent % of lag)
testspawn game from 100 to 500 ms (Butterflies again and drifters occasionally)
gmleVintagestory.GameContent.BlockEntityTransient (major impact when people explore) which we didnt do on stress-test.
entity-done-bh-harvestable (Around 10-15% of ticks..depends if people stay idle or explore..if they explore it goes quite more) and become one of the tier one source of overloading.

Tier 3 Culprits (Recurrent ones counting for about 10% of logticks overloading.
gmleVintagestory.GameContent.BlockEntityBerryBush(still above 100 ms-300ms sometimes) usually 10% of a tick.
gmleVintagestory.GameContent.BlockEntityFarmland Usually around 10% of a tick.

---------------------------------
On a side note here is some warnings and errors we found while going thru the server-main.txt posted above.

**Some Errors/Warnings Found:**
```
19.10.2021 07:48:53 [Event] Player Saraty got removed. Reason: Threw an exception at the server
19.10.2021 07:48:53 [Error] System.NullReferenceException: Object reference not set to an instance of an object
  at Vintagestory.GameContent.BlockMeal.SetContents (System.String recipeCode, Vintagestory.API.Common.ItemStack containerStack, Vintagestory.API.Common.ItemStack[] stacks, System.Single quantityServings) [0x0003b] in <a02030b72f6d4db7b3a1e2f8cb5fb657>:0 
  at Vintagestory.GameContent.BlockCookedContainerBase.ServeIntoBowlStack (Vintagestory.API.Common.ItemSlot bowlSlot, Vintagestory.API.Common.ItemSlot potslot, Vintagestory.API.Common.IWorldAccessor world) [0x00273] in <a02030b72f6d4db7b3a1e2f8cb5fb657>:0 
  at Vintagestory.GameContent.BlockFirepit.OnBlockInteractStart (Vintagestory.API.Common.IWorldAccessor world, Vintagestory.API.Common.IPlayer byPlayer, Vintagestory.API.Common.BlockSelection blockSel) [0x00279] in <a02030b72f6d4db7b3a1e2f8cb5fb657>:0 
  at Vintagestory.Server.ServerSystemBlockSimulation.HandleBlockInteract (_nSxSiZFWZ0rwDyGNmypOtEp4Igh packet, Vintagestory.Server.ConnectedClient client) [0x00196] in <c29bb51c062d406599d97705b6ec16ba>:0 
  at Vintagestory.Server.ServerSystemInventory.HandleHandInteraction (_nSxSiZFWZ0rwDyGNmypOtEp4Igh packet, Vintagestory.Server.ConnectedClient client) [0x00017] in <c29bb51c062d406599d97705b6ec16ba>:0 
  at Vintagestory.Server.ServerMain.HandleClientPacket (Vintagestory.Server.ConnectedClient client, System.Byte[] data) [0x0005a] in <c29bb51c062d406599d97705b6ec16ba>:0 
  at Vintagestory.Server.ServerMain.ProcessNetMessage (Vintagestory.Common.NetIncomingMessage msg, Vintagestory.Common.NetServer mainSocket) [0x00150] in <c29bb51c062d406599d97705b6ec16ba>:0 
---
21.10.2021 14:25:19 [Warning] Exception at client 242. Disconnecting client.
21.10.2021 14:25:19 [Event] Player Zonarius got removed. Reason: Threw an exception at the server
21.10.2021 14:25:19 [Error] System.NullReferenceException: Object reference not set to an instance of an object
  at Vintagestory.GameContent.BlockMeal.SetContents (System.String recipeCode, Vintagestory.API.Common.ItemStack containerStack, Vintagestory.API.Common.ItemStack[] stacks, System.Single quantityServings) [0x0003b] in <a02030b72f6d4db7b3a1e2f8cb5fb657>:0 
  at Vintagestory.GameContent.BlockCookedContainerBase.ServeIntoBowlStack (Vintagestory.API.Common.ItemSlot bowlSlot, Vintagestory.API.Common.ItemSlot potslot, Vintagestory.API.Common.IWorldAccessor world) [0x00273] in <a02030b72f6d4db7b3a1e2f8cb5fb657>:0 
  at Vintagestory.GameContent.BlockFirepit.OnBlockInteractStart (Vintagestory.API.Common.IWorldAccessor world, Vintagestory.API.Common.IPlayer byPlayer, Vintagestory.API.Common.BlockSelection blockSel) [0x00279] in <a02030b72f6d4db7b3a1e2f8cb5fb657>:0 
  at Vintagestory.Server.ServerSystemBlockSimulation.HandleBlockInteract (_nSxSiZFWZ0rwDyGNmypOtEp4Igh packet, Vintagestory.Server.ConnectedClient client) [0x00196] in <c29bb51c062d406599d97705b6ec16ba>:0 
  at Vintagestory.Server.ServerSystemInventory.HandleHandInteraction (_nSxSiZFWZ0rwDyGNmypOtEp4Igh packet, Vintagestory.Server.ConnectedClient client) [0x00017] in <c29bb51c062d406599d97705b6ec16ba>:0 
  at Vintagestory.Server.ServerMain.HandleClientPacket (Vintagestory.Server.ConnectedClient client, System.Byte[] data) [0x0005a] in <c29bb51c062d406599d97705b6ec16ba>:0 
  at Vintagestory.Server.ServerMain.ProcessNetMessage (Vintagestory.Common.NetIncomingMessage msg, Vintagestory.Common.NetServer mainSocket) [0x00150] in <c29bb51c062d406599d97705b6ec16ba>:0 
```
----------------------------------------
```
0.10.2021 03:40:48 [Warning] Unable to connect to auth server: State Error, Error msg 'Exception thrown when trying to connect. System.Reflection.TargetInvocationException: An exception occurred during the operation, making the result invalid.  Check InnerException for exception details. ---> System.Net.WebException: Error getting response stream (ReadDoneAsync2): ReceiveFailure
  at System.Net.WebResponseStream.InitReadAsync (System.Threading.CancellationToken cancellationToken) [0x000f3] in <9c6e2cb7ddd8473fa420642ddcf7ce48>:0 
  at System.Net.WebOperation.Run () [0x001d9] in <9c6e2cb7ddd8473fa420642ddcf7ce48>:0 
  at System.Net.WebCompletionSource`1[T].WaitForCompletion () [0x00094] in <9c6e2cb7ddd8473fa420642ddcf7ce48>:0 
  at System.Net.HttpWebRequest.RunWithTimeoutWorker[T] (System.Threading.Tasks.Task`1[TResult] workerTask, System.Int32 timeout, System.Action abort, System.Func`1[TResult] aborted, System.Threading.CancellationTokenSource cts) [0x000f8] in <9c6e2cb7ddd8473fa420642ddcf7ce48>:0 
  at System.Net.HttpWebRequest.EndGetResponse (System.IAsyncResult asyncResult) [0x00020] in <9c6e2cb7ddd8473fa420642ddcf7ce48>:0 
  at System.Net.WebClient.GetWebResponse (System.Net.WebRequest request, System.IAsyncResult result) [0x00000] in <9c6e2cb7ddd8473fa420642ddcf7ce48>:0 
  at System.Net.WebClient.GetWebResponseTaskAsync (System.Net.WebRequest request) [0x0008d] in <9c6e2cb7ddd8473fa420642ddcf7ce48>:0 
  at System.Net.WebClient.DownloadBitsAsync (System.Net.WebRequest request, System.IO.Stream writeStream, System.ComponentModel.AsyncOperation asyncOp, System.Action`3[T1,T2,T3] completionDelegate) [0x0008d] in <9c6e2cb7ddd8473fa420642ddcf7ce48>:0 
   --- End of inner exception stack trace ---
  at System.ComponentModel.AsyncCompletedEventArgs.RaiseExceptionIfNecessary () [0x0001d] in <9c6e2cb7ddd8473fa420642ddcf7ce48>:0 
  at System.Net.UploadValuesCompletedEventArgs.get_Result () [0x00000] in <9c6e2cb7ddd8473fa420642ddcf7ce48>:0 
  at Vintagestory.Common.MyWebClient.onComplete (System.Object sender, System.Net.UploadValuesCompletedEventArgs e) [0x00033] in <c29bb51c062d406599d97705b6ec16ba>:0 '
20.10.2021 03:40:48 [Notification] If this is your first server set-up, make sure the mono SSL Certificates are installed: https://wiki.vintagestory.at/index.php?title=Installing_the_game_on_Linux
```
------------------
```
24.10.2021 12:06:09 [Error] Exception thrown during handling event OnTrySpawnEntity: System.NullReferenceException: Object reference not set to an instance of an object
  at Vintagestory.GameContent.SystemTemporalStability.Event_OnTrySpawnEntity (Vintagestory.API.Common.Entities.EntityProperties& properties, Vintagestory.API.MathTools.Vec3d spawnPosition, System.Int64 herdId) [0x0003c] in <a02030b72f6d4db7b3a1e2f8cb5fb657>:0 
  at (wrapper delegate-invoke) <Module>.invoke_bool_EntityProperties&_Vec3d_long(Vintagestory.API.Common.Entities.EntityProperties&,Vintagestory.API.MathTools.Vec3d,long)
  at Vintagestory.Server.ServerEventManager.TriggerTrySpawnEntity (Vintagestory.API.Common.Entities.EntityProperties& properties, Vintagestory.API.MathTools.Vec3d position, System.Int64 herdId) [0x0000c] in <c29bb51c062d406599d97705b6ec16ba>:0 
--------
24.10.2021 12:06:09 [Error] Exception thrown during handling event OnTrySpawnEntity: System.NullReferenceException: Object reference not set to an instance of an object
  at Vintagestory.GameContent.SystemTemporalStability.Event_OnTrySpawnEntity (Vintagestory.API.Common.Entities.EntityProperties& properties, Vintagestory.API.MathTools.Vec3d spawnPosition, System.Int64 herdId) [0x0003c] in <a02030b72f6d4db7b3a1e2f8cb5fb657>:0 
  at (wrapper delegate-invoke) <Module>.invoke_bool_EntityProperties&_Vec3d_long(Vintagestory.API.Common.Entities.EntityProperties&,Vintagestory.API.MathTools.Vec3d,long)
  at Vintagestory.Server.ServerEventManager.TriggerTrySpawnEntity (Vintagestory.API.Common.Entities.EntityProperties& properties, Vintagestory.API.MathTools.Vec3d position, System.Int64 herdId) [0x0000c] in <c29bb51c062d406599d97705b6ec16ba>:0 
24.10.2021 12:06:09 [Error] Exception thrown during handling event OnTrySpawnEntity: System.NullReferenceException: Object reference not set to an instance of an object
  at Vintagestory.GameContent.SystemTemporalStability.Event_OnTrySpawnEntity (Vintagestory.API.Common.Entities.EntityProperties& properties, Vintagestory.API.MathTools.Vec3d spawnPosition, System.Int64 herdId) [0x0003c] in <a02030b72f6d4db7b3a1e2f8cb5fb657>:0 
  at (wrapper delegate-invoke) <Module>.invoke_bool_EntityProperties&_Vec3d_long(Vintagestory.API.Common.Entities.EntityProperties&,Vintagestory.API.MathTools.Vec3d,long)
  at Vintagestory.Server.ServerEventManager.TriggerTrySpawnEntity (Vintagestory.API.Common.Entities.EntityProperties& properties, Vintagestory.API.MathTools.Vec3d position, System.Int64 herdId) [0x0000c] in <c29bb51c062d406599d97705b6ec16ba>:0 
```
 -------------------
Those BETransient reports happen quite often i only put a few of them.
 23.10.2021 10:49:19 [Error] BETransient exiting at 484835, 112, 507267 cannot find block attributes. Will stop transient timer
23.10.2021 10:49:44 [Error] BETransient exiting at 484835, 112, 507267 cannot find block attributes. Will stop transient timer
2.10.2021 04:08:14 [Error] BETransient exiting at 478070, 113, 502159 cannot find block attributes. Will stop transient timer
23.10.2021 10:19:01 [Error] BETransient exiting at 484835, 112, 507267 cannot find block attributes. Will stop transient timer

### How to reproduce
We did a call to have as many people on at the same time in order to stress test the performance and provide more data to the developers.
### Expected behavior
Optimize the server code to reduce the overloads on the culprits identified above.
Utilize the server ressources more for RAM + CPU multi-thread in order to alleviate the overticks...
at 40 people we were roughly at only 9-14 gig of ram used while we have 32 gig available.
at 40 people only 2 core-threads were used vs all the ones we got available.

### Screenshots

![2021-10-24_15-15-44](https://user-images.githubusercontent.com/31166271/138764767-a278c196-9592-414e-8ec1-5511ad7b8bb5.png)

![130PM_CDT_38_Players](https://user-images.githubusercontent.com/31166271/138763580-5111243d-de5e-4aa5-9324-efe638fbb9d8.png)
![230_PM_CDT_37_Players](https://user-images.githubusercontent.com/31166271/138763583-05ce6e0b-ac55-4c84-9d20-bf627387172a.png)
![142PM_CDT_40_Players](https://user-images.githubusercontent.com/31166271/138763586-dd418552-8d91-4f0d-91a9-375259e71cc5.png)
![110pm_CDT](https://user-images.githubusercontent.com/31166271/138763587-8335eb41-b0a8-45b7-99ee-b6923200ebe0.png)
![120pm_CDT_35_players](https://user-images.githubusercontent.com/31166271/138763588-e7df0228-e453-42f6-9406-18cde851207d.png)
s

### Log
https://gist.githubusercontent.com/Aledark789/97a4af79721c8a7804a821e1a3581327/raw/b776b4c627339d5df9c035f8869e126c15b9f103/gistfile1.txt
right here @tyronx

**Comment by Aledark789** [comment]
> 
> 
> > (see log file link below)
> 
> Where is that link?

https://gist.githubusercontent.com/Aledark789/97a4af79721c8a7804a821e1a3581327/raw/b776b4c627339d5df9c035f8869e126c15b9f103/gistfile1.txt

**Comment by Th3Dilli** [comment]
with the changes made from reports in #2677 and others that were released with 1.18.5 this should be mostly solved

further we now have 1.18.8-rc.1 which is a net7 only version that also gives a huge performance boost

thank you aledark for all your efforts and continuous investigation into bugs, performance and reporting them
